lib/NoThrow.thy

```diff
@@ -0,0 +1,163 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+(* Lemmas about no_throw. Usually should have a conclusion "no_throw P m".
+   Includes some monad equations that have no_throw as a main assumption.  *)
+
+theory NoThrow
+  imports
+    WhileLoopRules
+    MonadEq_Lemmas
+begin
+
+(* Alternative definition of no_throw; easier to work with than unfolding validE. *)
+lemma no_throw_def': "no_throw P A = (\<forall>s. P s \<longrightarrow> (\<forall>(r, t) \<in> fst (A s). (\<exists>x. r = Inr x)))"
+  apply (clarsimp simp: no_throw_def validE_def2 split_def split: sum.splits)
+  done
+
+lemma no_throw_returnOk [simp]:
+  "no_throw P (returnOk a)"
+  apply (clarsimp simp: no_throw_def)
+  apply wp
+  done
+
+lemma no_throw_liftE [simp]:
+  "no_throw P (liftE x)"
+  apply (clarsimp simp: liftE_def no_throw_def validE_def)
+  apply (wp | simp)+
+  done
+
+lemma no_throw_bindE: "\<lbrakk> no_throw A X; \<And>a. no_throw B (Y a); \<lbrace> A \<rbrace> X \<lbrace> \<lambda>_. B \<rbrace>,\<lbrace> \<lambda>_ _. True \<rbrace> \<rbrakk> \<Longrightarrow> no_throw A (X >>=E Y)"
+  apply atomize
+  apply (clarsimp simp: no_throw_def)
+  apply (rule seqE [rotated])
+   apply force
+  apply (rule  hoare_validE_cases)
+   apply simp
+  apply simp
+  done
+
+lemma no_throw_bindE_simple: "\<lbrakk> no_throw \<top> L; \<And>x. no_throw \<top> (R x) \<rbrakk> \<Longrightarrow> no_throw \<top> (L >>=E R)"
+  apply (erule no_throw_bindE)
+   apply assumption
+  apply wp
+  done
+
+lemma no_throw_handleE_simple:
+  notes hoare_vcg_prop[wp del]
+  shows "\<lbrakk> \<And>x. no_throw \<top> L \<or> no_throw \<top> (R x) \<rbrakk> \<Longrightarrow> no_throw \<top> (L <handle> R)"
+  apply (clarsimp simp: no_throw_def)
+  apply atomize
+  apply clarsimp
+  apply (erule disjE)
+   apply (wpsimp wp: hoare_vcg_prop[where f="R x" for x])
+    apply assumption
+   apply simp
+  apply (rule handleE_wp)
+   apply (erule_tac x=x in allE)
+   apply assumption
+  apply wp
+  done
+
+lemma no_throw_handle2: "\<lbrakk> \<And>a. no_throw Y (B a); \<lbrace> X \<rbrace> A \<lbrace> \<lambda>_ _. True \<rbrace>,\<lbrace> \<lambda>_. Y \<rbrace> \<rbrakk> \<Longrightarrow> no_throw X (A <handle2> B)"
+  apply atomize
+  apply (clarsimp simp: no_throw_def' handleE'_def bind_def)
+  apply (clarsimp simp: validE_def valid_def)
+  apply (erule allE, erule (1) impE)
+  apply (drule (1) bspec)
+  apply (clarsimp simp: return_def split: sum.splits)
+  apply force
+  done
+
+lemma no_throw_handle: "\<lbrakk> \<And>a. no_throw Y (B a); \<lbrace> X \<rbrace> A \<lbrace> \<lambda>_ _. True \<rbrace>,\<lbrace> \<lambda>_. Y \<rbrace> \<rbrakk> \<Longrightarrow> no_throw X (A <handle> B)"
```

Comment:
this file doesn't have the greatest style, but I know it was copy/pasted... hmm

Reply:
Yes, these are all just moves, I've tried hard not to touch style or other cleanup. We should do those separately in this case, otherwise this gets too big.

Reply:
I caved in and cleaned up the worst of it..